### DISCUSSION TITLE
Help with debugger

### DISCUSSION THREAD

**Discussion by nelgin** [dosemu2/dosemu2#2247] [General]
Hi all,

Not sure if this is possible...

I want to find where some text is displayed in a program and considered trying to find the memory location of the text and then see if I can set a breakpoint when something tries to call that memory location. Does that even make sense? Is such a thing possible? Should I be looking at a different approach? I would obviously want to write the changes back to the executable file once I am done.

Rather than using dosdebug, are there other options I should consider for this particular task?

**Comment by ecm-pushbx** [comment]
Do you know whether the text is displayed using DOS / int 21h, or ROM-BIOS int 10h, or by directly writing to the text mode video memory? You can set breakpoints fairly easily for the first in lDebug, eg:

```
bp new ptr ri21p when (ah == 09 || ah == 40) && value linear ds:dx in from linear SEGMENT:OFFSET length LENGTH
```

Replace the allcaps keywords by your text's location in memory. This example would work if the text is displayed using the int 21h functions 09h or 40h.

It's more difficult if int 21h function 02h or int 10h function 0Eh or direct buffer writing are used. In this case you may have luck using lDebug's access variables, but this requires tracing the program which causes a huge slowdown. Example:

```
tp CCCCC while value from linear SEGMENT:OFFSET length LENGTH in reading silent 4
```

If this succeeds it will stop before running the first instruction that reads from your string.

**Reply by ecm-pushbx** [reply]
Oops. You need `tp CCCCC while ! value ...` of course.

**Comment by nelgin** [comment]
I think it's using int 10h but will have to check when I'm back on the computer.  But used iDebug and just heard about it tonight.  Will track it down.

Thanks

**Reply by ecm-pushbx** [reply]
It's lDebug with a small L. Manual and downloads (release or current daily build) are available at https://pushbx.org/ecm/web/#projects-ldebug

**Comment by nelgin** [comment]
What information you gave me helped a lot, thank you, though I didn't use ldebug in the end but I went with, care I say it, dosbox-x with its debugger. I managed to determine it's using int 21 with ah = 0x06. 

Looks like it's writing character by character to the screen.

Now I think I might have what I need. Thank you.
